File: src/classes/RealtimeData.ts
import Utils from "../utils/Utils";
import { IRealtimeData } from "../entities/IRealtimeData";
import { ICombinedRealtimeData } from "../entities/ICombinedRealtimeData";
import Logger from "../utils/Logger";
import CombinedRealtimeData from "./CombinedRealtimeData";

export class RealtimeData {
  private readonly _realtimeData: IRealtimeData[];
  constructor(realtimeData: IRealtimeData[]) {
    this._realtimeData = realtimeData;
  }
  get realtimeData() {
    return this._realtimeData;
  }

  getContainingNamespaces() {
    return new Set(this._realtimeData.map((r) => r.namespace));
  }

  toCombinedRealtimeData() {
    const uniqueNameMapping = new Map<string, IRealtimeData[]>();
    this._realtimeData.forEach((r) => {
      const id = r.uniqueEndpointName;
      uniqueNameMapping.set(id, (uniqueNameMapping.get(id) || []).concat([r]));
    });

    const combined = [...uniqueNameMapping.values()]
      .map((group): ICombinedRealtimeData[] => {
        const statusMap = new Map<string, IRealtimeData[]>();
        group.forEach((r) => {
          statusMap.set(r.status, (statusMap.get(r.status) || []).concat([r]));
        });
        const sample = group[0];
        const baseSample = {
          uniqueServiceName: sample.uniqueServiceName,
          uniqueEndpointName: sample.uniqueEndpointName,
          service: sample.service,
          namespace: sample.namespace,
          version: sample.version,
          method: sample.method,
        };

        const combinedSubGroup = [...statusMap.entries()].map(
          ([status, subGroup]): ICombinedRealtimeData => {
            const combined = subGroup.reduce((prev, curr) => {
              prev.latency += curr.latency;
              prev.requestBody = Utils.MergeStringBody(
                prev.requestBody,
                curr.requestBody
              );
              prev.responseBody = Utils.MergeStringBody(
                prev.responseBody,
                curr.responseBody
              );
              prev.timestamp =
                prev.timestamp > curr.timestamp
                  ? prev.timestamp
                  : curr.timestamp;
              if (prev.replica && curr.replica) prev.replica += curr.replica;
              return prev;
            });
            const { requestBody, requestSchema, responseBody, responseSchema } =
              this.parseRequestResponseBody(combined);

            return {
              ...baseSample,
              status,
              combined: subGroup.length,
              requestBody,
              requestSchema,
              responseBody,
              responseSchema,
              avgLatency: combined.latency / subGroup.length,
              avgReplica: combined.replica
                ? combined.replica / subGroup.length
                : undefined,
              latestTimestamp: combined.timestamp,
              latencies: subGroup.map((r) => r.latency),
            };
          }
        );
        return combinedSubGroup;
      })
      .flat();

    return new CombinedRealtimeData(combined);
  }

  private parseRequestResponseBody(data: IRealtimeData) {
    let requestBody: any | undefined;
    let requestSchema: string | undefined;
    let responseBody: any | undefined;
    let responseSchema: string | undefined;
    if (data.requestContentType === "application/json") {
      try {
        requestBody = JSON.parse(data.requestBody!);
        requestSchema = Utils.ObjectToInterfaceString(requestBody);
      } catch (e) {
        Logger.verbose(`Not a JSON, skipping: [${data.requestBody}]`);
        requestBody = requestSchema = undefined;
      }
    }
    if (data.responseContentType === "application/json") {
      try {
        responseBody = JSON.parse(data.responseBody!);
        responseSchema = Utils.ObjectToInterfaceString(responseBody);
      } catch (e) {
        Logger.verbose(`Not a JSON, skipping: [${data.responseBody}]`);
        responseBody = responseSchema = undefined;
      }
    }
    return {
      requestBody,
      requestSchema,
      responseBody,
      responseSchema,
    };
  }
}

File: kmamiz_data_processor/src/data/realtime_data.rs
use std::collections::HashMap;

use crate::json_utils;

use super::{
    combined_realtime_data::{CombinedLatency, CombinedRealtimeData},
    request_type::RequestType,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct RealtimeData {
    pub unique_service_name: String,
    pub unique_endpoint_name: String,
    pub timestamp: i64,
    pub method: RequestType,
    pub service: String,
    pub namespace: String,
    pub version: String,
    pub latency: u64,
    pub status: String,
    pub request_body: Option<String>,
    pub request_content_type: Option<String>,
    pub response_body: Option<String>,
    pub response_content_type: Option<String>,
    pub replica: Option<u32>,
}

impl RealtimeData {
    pub fn combine(data: Vec<RealtimeData>) -> Vec<CombinedRealtimeData> {
        let mut name_mapping = HashMap::new();
        data.into_iter().for_each(|d| {
            let id = format!(
                "{}\t{}\t{}\t{}",
                d.unique_endpoint_name,
                d.status,
                d.request_content_type.clone().unwrap_or_default(),
                d.response_content_type.clone().unwrap_or_default()
            );
            let entry = name_mapping.entry(id).or_insert(vec![]);
            entry.push(d);
        });

        name_mapping
            .into_values()
            .map(|group| {
                let combined = group.len() as f64;
                let sample = group[0].clone();

                let mut total_latency = 0;
                let mut total_replicas = 0;
                let mut latest_timestamp = 0;
                let mut request_body = vec![];
                let mut response_body = vec![];
                let mut div_base = 0;
                for data in group.into_iter() {
                    total_latency += data.latency;
                    if let Some(body) = data.request_body {
                        request_body.push(body);
                    }
                    if let Some(body) = data.response_body {
                        response_body.push(body);
                    }
                    latest_timestamp = latest_timestamp.max(data.timestamp);
                    div_base += data.latency.pow(2);
                    if let Some(replica) = data.replica {
                        total_replicas += replica;
                    }
                }
                let div_base = RealtimeData::to_precise(div_base as f64);
                let mean = RealtimeData::to_precise(total_latency as f64 / combined);
                let cv_top = (div_base / combined - mean.powi(2)).sqrt();

                let cv = if cv_top.is_normal() {
                    RealtimeData::to_precise(cv_top / mean)
                } else {
                    0.0
                };
                let latency = CombinedLatency { mean, div_base, cv };

                let request_body = Self::process_body(request_body);
                let response_body = Self::process_body(response_body);

                CombinedRealtimeData {
                    unique_service_name: sample.unique_service_name,
                    unique_endpoint_name: sample.unique_endpoint_name,
                    service: sample.service,
                    namespace: sample.namespace,
                    version: sample.version,
                    method: sample.method,
                    status: sample.status,
                    request_content_type: sample.request_content_type,
                    response_content_type: sample.response_content_type,
                    combined: combined as usize,
                    latency,
                    latest_timestamp,
                    request_body: serde_json::to_string(&request_body).ok(),
                    response_body: serde_json::to_string(&response_body).ok(),
                    request_schema: Some(json_utils::to_types(request_body)),
                    response_schema: Some(json_utils::to_types(response_body)),
                    avg_replica: total_replicas as f64 / combined,
                    _id: None,
                }
            })
            .collect()
    }

    fn process_body(samples: Vec<String>) -> Value {
        let samples = samples
            .into_iter()
            .filter_map(|req| serde_json::from_str(&req).ok())
            .collect::<Vec<Value>>();
        json_utils::merge(samples)
    }

    fn to_precise(num: f64) -> f64 {
        ((num + f64::EPSILON) * 1e14).round() / 1e14
    }
}
